### ISSUE TITLE
Default value not applied when it's defined through $ref

### ISSUE THREAD

**Issue by acelot** [justinrainbow/json-schema#473]
**Schema (YML):**
```yml
definitions:
  limit:
    type: integer
    minimum: 1
    maximum: 100
    default: 10

test:
  type: object
  properties:
    query:
      type: string
    limit:
      $ref: "#/definitions/limit"
    offset:
      type: integer
      default: 0
```

**JSON:**
```json
{
  "query": "test"
}
```

**Validation:**
```php
$validator->validate($request, $schema, Constraint::CHECK_MODE_APPLY_DEFAULTS);
```

**Request after validation:**
```json
{
  "query": "test",
  "offset": 0
}
```

It seems validator doesn't apply default value in $refs.
Version: 5.2.6
PHP: 7.1

**Comment by guillemcanal** [comment]
I'm having the same issue, I'll investigate

**Comment by erayd** [comment]
@acelot My apologies - I missed this issue!

Looking at your schema, it appears you have most of it defined under #/test, which is not a valid location for a subschema. *Any* data instance would be considered valid under the schema you've provided, as there aren't actually any valid constraints in scope. This also means that the default value is out of scope. However, you've also posted data that has been modified by the schema.

Could you please post the actual schema that's being passed to the validator? Or if that *is* the actual schema, let me know, because it means there is actually a different bug present.

**Comment by acelot** [comment]
@erayd all my schemes (and definitions) are stored in single YAML config file. Actual validator looks like this:

```php
$schema = <<<'JSON'
{
  "test": {
    "type": "object", 
    "properties": {
      "query": {
        "type": "string"
      }, 
      "limit": {
        "$ref": "#/definitions/limit"
      }, 
      "offset": {
        "default": 0, 
        "type": "integer"
      }
    }
  }, 
  "definitions": {
    "limit": {
      "default": 10, 
      "minimum": 1, 
      "type": "integer", 
      "maximum": 100
    }
  }
}
JSON;

$body = <<<'JSON'
{
  "query": "test"
}
JSON;

$schema = json_decode($schema);
$body = json_decode($body);

$schemaStorage = new SchemaStorage();
$schemaStorage->addSchema('file://schema', $schema);
$validator = new Validator(new Factory($schemaStorage));
$validator->validate($body, $schema, Constraint::CHECK_MODE_NORMAL | Constraint::CHECK_MODE_APPLY_DEFAULTS);
```

**Comment by erayd** [comment]
@acelot There is definitely some user error here. The bug you mention is a real one, but the example you have posted is not a valid way to replicate it, and cannot have triggered the bug.

 * The effective schema you're actually validating against is `{}`<sup>1</sup>.
 * `{}` contains no constraint keywords, and does not set any defaults.
 * The correct result is for `$body` to be considered valid, and not altered.
 * v5.2.6 of this library behaves correctly according to the above.

<sup>1</sup> It looks like you have tried to define your schema inside of `#/test`. Don't do this; it's not a valid location - and as a result, the whole thing gets ignored, as per the spec. The contents of `#/test` actually belongs in the schema root.

The schema below is the correct way to do this, and also replicates the not-setting-referenced-defaults bug:
```JSON
{
  "type": "object", 
  "properties": {
    "query": {
      "type": "string"
    }, 
    "limit": {
      "$ref": "#/definitions/limit"
    }, 
    "offset": {
      "default": 0, 
      "type": "integer"
    }
  },
  "definitions": {
    "limit": {
      "default": 10, 
      "minimum": 1, 
      "type": "integer", 
      "maximum": 100
    }
  }
}
```

I am looking into the bug now - stand by :-).

**Comment by acelot** [comment]
Yeah, it's my fault, in fact, there is no key `test` at root of schema JSON. I have update my comment.

**Comment by erayd** [comment]
@acelot Fixed in #473. Once it's been reviewed & merged, I'll backport it for v5.2.7.

The root cause was checking for a default value on a non-dereferenced schema. The fix dereferences the property / item definition first.

**Comment by erayd** [comment]
Backported for 5.2.7 in #495.
